FORKING_FLUTTER.md

```diff
@@ -0,0 +1,110 @@
+rendering the UI, handling input, and communicating with the host.
+
+We forked this code to add the ability to have release versions of the Flutter
+engine be able to load new code from Shorebird's servers.
+
+At time of writing, Shorebird's fork is based on Flutter 3.7.8.  You can see
+our engine changes here:
+https://github.com/flutter/engine/compare/3.7.8...shorebirdtech:engine:stable_codepush
+
+### flutter/flutter
+
+The flutter/flutter repo contains the Dart code that runs on the device as well
+as the `flutter` tool that is used to build and run Flutter apps.
+
+We initially did not fork this code.  And still don't really want to fork
+this code, but in order to deliver a modified engine w/o affecting other
+Flutter installations, we needed to be able to change the *version* of the
+engine that the `flutter` tool downloads.
+
+Our one fork is to change bin/internal/engine.version to point to our
+engine version.  You can see our changes here:
+https://github.com/flutter/flutter/compare/3.7.8...shorebirdtech:flutter:stable_codepush
+
+
+### flutter/buildroot
+
+The buildroot repo contains the build scripts that are used to build the
+Flutter engine for various platforms.  It's separate from flutter/engine in
+order to share code and configuration with the Fuchsia build system.
+
+We also didn't want to fork this code.  However we need to for now in order
+to integrate our updater code.  Our updater code:
+https://github.com/shorebirdtech/updater
+is a Rust library which we link into the engine.  The way we do that is via
+a C-API on a static library (libupdater.a).  The default flags for linking
+for the Flutter engine hide all symbols from linked static libraries.  We
+need to be able to expose the shorebird_* symbols from libupdater.a up through
+FFI to the Dart code.  We did that my making one change to buildroot and then
+a second change to the engine to place the symbols on the allow-list.
+
+Our one change:
+https://github.com/shorebirdtech/buildroot/commit/7383548fa2306b5d53979ac5e9d176b35258811b
+
+
+## Vendoring our fork
+
+When you install Shorebird, it installs Flutter and Dart from our fork.  These
+are currently not exposed on the user's path, rather just private copies
+that Shorebird will use when building your app.
+
+This was necessary to avoid conflicts with other Flutter installations on the
+user's machine.  Specifically, the way that Flutter downloads artifacts is
+based on the version of the engine.  If we were to use the same version of the
+engine as the user's Flutter installation, then we would overwrite the user's
+engine artifacts.
+
+We deliver our artifacts to this fork of Flutter with two ways.  First is we
+change the version of the engine in the `flutter` tool.  Second is we pass
+FLUTTER_STORAGE_BASE_URL set to download.shorebird.dev (instead of
+download.flutter.io) when calling our vended copy of the `flutter` tool.
+
+Currently this means `shorebird` will not work in an environment where the
```

Comment:
Unless we just modify the flutter_tool to use `download.shorebird.dev` instead of relying on the environment variable 🤷

Reply:
Or we detect that they have it set and we let it override, etc.  There are choices we can make here once we have someone to bounce these off of.

We also might not need this if we go the CDN route.